chore: Update Service exposition concepts page (#760)
stackabletech/documentation: modules/concepts/pages/service-exposition.adoc

:listenerclass: xref:listener-operator:listenerclass.adoc
:description: Explore how Stackable uses listener-operator to expose Services.

Data products expose interfaces to the outside world.
These interfaces (whether UIs, or APIs) can be accessed by other products or by end users.
Clients accessing the interfaces can run inside or outside of the same Kubernetes cluster.
For example, xref:zookeeper:index.adoc[Apache ZooKeeper] is a dependency for other products, and it usually needs to be accessible only from within Kubernetes, while xref:superset:index.adoc[Apache Superset] is a data analysis product for end users and therefore needs to be accessible from outside the Kubernetes cluster.
Users connecting to Superset can be restricted within the local company network, or they can connect over the internet depending on the company security policies and demands.
This page gives an overview over the different options for service exposition, when to choose which option and how these options are configured.

== Motivation

Service exposition is such a complicated topic, that Stackable has build it's own operator for that: {listener-operator}[].

> The following section explains the motivation why we wrote such an operator over just using plain regular Kubernetes Services.

Reviewer: ```suggestion
The following section explains the motivation behind implementing such an operator instead of using plain regular Kubernetes Services.
```

> === Tools advertising their address

Reviewer: ```suggestion
=== Tools advertising their addresses
```

> Some tools need to know how they are externally reachable.

Reviewer: ```suggestion
Some products require information about their external accessibility.
```

> This is e.g. important for HDFS, where the namenode keeps track of which datanode serves which block or Kafka (used for client bootstrapping).

Reviewer: ```suggestion
This is e.g. important for HDFS, where the namenode keeps track of which datanode serves which block. Another case is Kafka, where it is required for client bootstrapping.
```

> A HDFS client asks the namenode "I want to read block 42, who is serving that?", the namenode responds with "block 42 is served by <ip or hostname of some datanode>".

Reviewer: ```suggestion
A common use case is an HDFS client connecting to a namenode in order to read block 42. Therefore, the namenode needs to know which datanode is serving block 42. The namenode then responds with the IP or hostname of the datanode containing that block 42.
```

For that to work, the datanode needs to know it's external address on startup and tell it the namenode.

> (And yes, we needed to patch Hadoop source-code for that ;))

Reviewer: I like comments like that, loosens up the reading flow etc, but not sure about the smiley :)

The {listener-operator}[listener-operator] runs as CSI driver (same as the {secret-operator}[secret-operator]) and places files inside the CSI volume, which tell the tool how it is reachable.

=== Integration with {secret-operator}[secret-operator]

If a tool is secured using TLS or Kerberos, it does not only need to be reachable via the determined address, it also needs a TLS certificate/keytab issued on the determined address.
{secret-operator}[secret-operator] integrated with to {listener-operator}[listener-operator], so that the platform takes care of provisioning certificates with the correct addresses (in the form of SAN entries).

== {listenerclass}[ListenerClasses]

A {listenerclass}[] describes how a product should be exposed.

> Please read on {listenerclass}[it's documentation] before continuing on this page.

Reviewer: ```suggestion
Please read on {listenerclass}[its documentation] before continuing on this page.
```

As a quick reminder, the platform ships with 3 default {listenerclass}[ListenerClasses]:

`cluster-internal`:: Used for listeners that are only accessible internally from the cluster. For example: communication between ZooKeeper nodes.
`external-unstable`:: Used for listeners that are accessible from outside the cluster, but which do not require a stable address. For example: individual Kafka brokers.
`external-stable`:: Used for listeners that are accessible from outside the cluster, and do require a stable address. For example: Kafka bootstrap.

Keep in mind that you are not restricted to this list, you can configure your own custom {listenerclass}[ListenerClasses].

> == Configuring the ListenerClass for a stacklet

Reviewer: ```suggestion
== Configuring the ListenerClass for a Stacklet
```

> We integrated {listener-operator}[listener-operator] into most of our products, currently only xref:opa:index.adoc[] and xref:spark-k8s:index.adoc[] are not using {listener-operator}[listener-operator].

Reviewer: ```suggestion
The {listener-operator}[listener-operator] is integrated into most of the Stackable products, currently only xref:opa:index.adoc[] and xref:spark-k8s:index.adoc[] are not using {listener-operator}[listener-operator].
```

> Most of the tools configure the {listenerclass}[] at the role level as follows:

Reviewer: Could we highlight the differences to products that do not do that? e.g. role group level listener etc?

Author: WDYT a886b67?
